File: server/src/sb.ts
import type { Db } from './db.ts'
import { runQuery, type FilterOp, type QueryRequest } from './query.ts'

/**
 * Supabase 클라이언트와 같은 모양의 빌더.
 *
 * Edge Function 들이 `admin.from('orders').select('*').eq('id', x).maybeSingle()`
 * 형태로 쓰여 있다. 그 코드를 그대로 옮기기 위해 같은 문법을 게이트웨이 위에 얹는다.
 * 반환도 Supabase 와 같은 { data, error } 다.
 */

export interface Result<T = any> { data: T; error: { message: string } | null }

class Builder implements PromiseLike<Result> {
  private request: QueryRequest
  private db: Db

  constructor(db: Db, table: string) {
    this.db = db
    this.request = { table, op: 'select' }
  }

  select(columns = '*') { this.request.select = columns; if (this.request.op !== 'select') this.request.returning = true; return this }
  insert(values: any) { this.request.op = 'insert'; this.request.values = values; return this }
  upsert(values: any, options?: { onConflict?: string }) {
    this.request.op = 'upsert'; this.request.values = values
    if (options?.onConflict) this.request.onConflict = options.onConflict
    return this
  }
  update(values: any) { this.request.op = 'update'; this.request.values = values; return this }
  delete() { this.request.op = 'delete'; return this }

  private filter(column: string, op: FilterOp, value: unknown) {
    ;(this.request.filters ??= []).push({ column, op, value })
    return this
  }
  eq(column: string, value: unknown) { return this.filter(column, 'eq', value) }
  neq(column: string, value: unknown) { return this.filter(column, 'neq', value) }
  in(column: string, value: unknown[]) { return this.filter(column, 'in', value) }
  gte(column: string, value: unknown) { return this.filter(column, 'gte', value) }
  lte(column: string, value: unknown) { return this.filter(column, 'lte', value) }
  is(column: string, value: unknown) { return this.filter(column, 'is', value) }

  order(column: string, options?: { ascending?: boolean }) {
    ;(this.request.order ??= []).push({ column, ascending: options?.ascending ?? true })
    return this
  }
  limit(n: number) { this.request.limit = n; return this }
  single() { this.request.single = 'one'; return this }
  maybeSingle() { this.request.single = 'maybe'; return this }

  async run(): Promise<Result> {
    try {
      const { data } = await runQuery(this.db, this.request)
      return { data, error: null }
    } catch (error) {
      return { data: null, error: { message: error instanceof Error ? error.message : String(error) } }
    }
  }

  then<A, B>(onfulfilled?: ((value: Result) => A | PromiseLike<A>) | null,
             onrejected?: ((reason: unknown) => B | PromiseLike<B>) | null): PromiseLike<A | B> {
    return this.run().then(onfulfilled, onrejected)
  }
}

export function sb(db: Db) {
  return { from: (table: string) => new Builder(db, table) }
}

File: server-py/app/sb.py
"""
Supabase 클라이언트와 같은 모양의 빌더.

Edge Function 들이 `admin.from_('orders').select('*').eq('id', x).maybe_single()`
형태로 쓰여 있다. 그 코드를 그대로 옮기기 위해 같은 문법을 게이트웨이 위에 얹는다.
반환도 Supabase 와 같은 (data, error) 다.

체이닝 끝에서 await 하면 실행된다 — __await__ 를 구현해 뒀다.
"""
from dataclasses import dataclass
from typing import Any, Generator

import asyncpg

from .query import run_query


@dataclass
class Result:
    data: Any
    error: dict | None


class Builder:
    def __init__(self, conn: asyncpg.Connection, table: str):
        self._conn = conn
        self._request: dict[str, Any] = {"table": table, "op": "select"}

    def select(self, columns: str = "*") -> "Builder":
        self._request["select"] = columns
        if self._request["op"] != "select":
            self._request["returning"] = True
        return self

    def insert(self, values: Any) -> "Builder":
        self._request.update(op="insert", values=values)
        return self

    def upsert(self, values: Any, on_conflict: str | None = None) -> "Builder":
        self._request.update(op="upsert", values=values)
        if on_conflict:
            self._request["onConflict"] = on_conflict
        return self

    def update(self, values: Any) -> "Builder":
        self._request.update(op="update", values=values)
        return self

    def delete(self) -> "Builder":
        self._request["op"] = "delete"
        return self

    def _filter(self, column: str, op: str, value: Any) -> "Builder":
        self._request.setdefault("filters", []).append({"column": column, "op": op, "value": value})
        return self

    def eq(self, column: str, value: Any) -> "Builder":
        return self._filter(column, "eq", value)

    def neq(self, column: str, value: Any) -> "Builder":
        return self._filter(column, "neq", value)

    def in_(self, column: str, value: list) -> "Builder":
        return self._filter(column, "in", value)

    def gte(self, column: str, value: Any) -> "Builder":
        return self._filter(column, "gte", value)

    def lte(self, column: str, value: Any) -> "Builder":
        return self._filter(column, "lte", value)

    def is_(self, column: str, value: Any) -> "Builder":
        return self._filter(column, "is", value)

    def order(self, column: str, ascending: bool = True) -> "Builder":
        self._request.setdefault("order", []).append({"column": column, "ascending": ascending})
        return self

    def limit(self, n: int) -> "Builder":
        self._request["limit"] = n
        return self

    def single(self) -> "Builder":
        self._request["single"] = "one"
        return self

    def maybe_single(self) -> "Builder":
        self._request["single"] = "maybe"
        return self

    async def run(self) -> Result:
        try:
            result = await run_query(self._conn, self._request)
            return Result(result.data, None)
        except Exception as error:  # noqa: BLE001 — Supabase 와 같이 오류를 값으로 돌려준다
            return Result(None, {"message": str(error)})

    def __await__(self) -> Generator[Any, None, Result]:
        return self.run().__await__()


class Sb:
    def __init__(self, conn: asyncpg.Connection):
        self._conn = conn

    def from_(self, table: str) -> Builder:
        return Builder(self._conn, table)


def sb(conn: asyncpg.Connection) -> Sb:
    return Sb(conn)
